File: src/com/jetbrains/idear/asr/ListeningState.java
package com.jetbrains.idear.asr;

import java.util.concurrent.atomic.AtomicReference;

/**
 * Created by breandan on 10/23/2015.
 */
public class ListeningState {
    private static final AtomicReference<Status> status = new AtomicReference<>(Status.INIT);

    public enum Status {
        INIT,
        ACTIVE,
        STANDBY,
        TERMINATED
    }

    private static Status setStatus(Status s) {
        return status.getAndSet(s);
    }

    public static Status getStatus() {
        return status.get();
    }

    public static boolean isTerminated() {
        return getStatus() == Status.TERMINATED;
    }

    public static boolean isInit() {
        return getStatus() == Status.INIT;
    }

    public static boolean isActive() {
        return getStatus() == Status.ACTIVE;
    }

    public static boolean standBy() {
        return Status.STANDBY == setStatus(Status.STANDBY);
    }

    public static boolean activate() {
        return Status.ACTIVE == setStatus(Status.ACTIVE);
    }
}

File: src/main/java/com/jetbrains/idear/asr/ListeningState.kt
package com.jetbrains.idear.asr

import com.jetbrains.idear.asr.ListeningState.Status.*
import java.util.concurrent.atomic.AtomicReference

/**
 * Created by breandan on 10/23/2015.
 */
object ListeningState {
    private val status = AtomicReference(INIT)

    enum class Status {
        INIT,
        ACTIVE,
        STANDBY,
        TERMINATED
    }

    private fun setStatus(s: Status): Status {
        return status.getAndSet(s)
    }

    fun getStatus(): Status {
        return status.get()
    }

    val isTerminated: Boolean
        get() = getStatus() == TERMINATED

    val isInit: Boolean
        get() = getStatus() == INIT

    val isActive: Boolean
        get() = getStatus() == ACTIVE

    fun standBy(): Boolean {
        return STANDBY == setStatus(STANDBY)
    }

    fun activate(): Boolean {
        return ACTIVE == setStatus(ACTIVE)
    }
}
